### PR TITLE
NOISSUE - Set keep alive to TCP connections

### PR DESCRIPTION
**Pull request by dusanb94**
Set KeepAlive for incoming and outgoing TCP connections

Signed-off-by: dusanb94 <dusan.borovcanin@mainflux.com>

### CODE REVIEW

File: pkg/mqtt/mqtt.go [hunk 1]
```diff
@@ -44,11 +50,20 @@ func (p Proxy) handle(inbound net.Conn) {
...
+		p.logger.Error("Cannot connect to remote broker " + p.target + " due to: " + err.Error())
 		return
 	}
 	defer p.close(outbound)
 
+	if err := setKeepAlive(inbound, p.keepAlive); err != nil {
+		p.logger.Error("Client conenction error: " + err.Error())
+		return
+	}
+	if err := setKeepAlive(inbound, p.keepAlive); err != nil {
```
Comment (drasko) [inline pkg/mqtt/mqtt.go:62]: Looks like `setKeepAlive()` was made 2 times on `inbound`.

Do we have to handle inbound at all - it is more role of the client (in this case Envoy proxy probably, and I doubt that they do not do it right)?

File: pkg/mqtt/mqtt.go [hunk 2]
```diff
@@ -76,3 +91,11 @@ func (p Proxy) close(conn net.Conn) {
 		p.logger.Warn(fmt.Sprintf("Error closing connection %s", err.Error()))
 	}
 }
+
+func setKeepAlive(conn net.Conn, period time.Duration) error {
+	tcpConn, ok := conn.(*net.TCPConn)
+	if !ok {
+		return errNotTCPConn
+	}
+	return tcpConn.SetKeepAlivePeriod(period)
```
Comment (drasko) [inline pkg/mqtt/mqtt.go:100]: Do we have to "enable" keepalive by setting it to `true` as well (not only setting period), like [here](https://github.com/inetaf/tcpproxy/blob/master/tcpproxy.go#L377)?

Comment (dusanb94) [reply pkg/mqtt/mqtt.go:100]: We do not, but it makes more sense. I'll add the flag.

File: pkg/mqtt/mqtt.go (hunk 1 above)
Comment (dusanb94) [reply pkg/mqtt/mqtt.go:62]: I will take a look. It will probably be removed because I agree - the client should take care of this.

Comment (nmarcetic) [reply pkg/mqtt/mqtt.go]: If we will handle only the outbound connection, you can just use [DialTimeout()](https://golang.org/pkg/net/#DialTimeout) and simplify everything.

Comment (drasko) [reply pkg/mqtt/mqtt.go]: Isn't that used only for timeout during dial (connection establishment), and not as a keepalive (connection heartbeat during the life of connection)?

Comment (dusanb94) [reply pkg/mqtt/mqtt.go]: Yes, it is. However, Dialer itself can set `KeepAlive` timeout for each Connection it creates, which is how it works now. The same goes for Listener, so there is no need to explicitly set the timeout for each connection.

File: pkg/mqtt/mqtt.go (hunk 2 above)
Comment (dusanb94) [reply pkg/mqtt/mqtt.go]: This is done by Dialer and Listener now, as well.

File: cmd/main.go [hunk 3]
```diff
@@ -101,6 +111,15 @@ func env(key, fallback string) string {
 }
 
 func loadConfig() config {
+	ka, err := strconv.ParseBool(mainflux.Env(envKeepAlive, defKeepAlive))
+	if err != nil {
+		log.Fatalf("Invalid value passed for %s\n", envKeepAlive)
+	}
+	kaPeriod, err := strconv.ParseInt(mainflux.Env(envKeepAlivePeriod, defKeepAlivePeriod), 10, 64)
+	if err != nil {
+		log.Fatalf("Invalid %s value: %s", envKeepAlivePeriod, err.Error())
```
Comment (manuio) [inline cmd/main.go:120]: Maybe better `Invalid period value`?

File: cmd/main.go [hunk 4]
```diff
@@ -101,6 +111,15 @@ func env(key, fallback string) string {
 }
 
 func loadConfig() config {
+	ka, err := strconv.ParseBool(mainflux.Env(envKeepAlive, defKeepAlive))
+	if err != nil {
+		log.Fatalf("Invalid value passed for %s\n", envKeepAlive)
```
Comment (manuio) [inline cmd/main.go:116]: `Invalid keep alive value passed` ?

**Pull request merged**